Sprint 13

Fen1x12 opened the pull request.

Dedtop committed:
изменил название файлов

Dedtop committed:
написал код для экрана и убрал лишние комментария

Dedtop committed:
1

Dedtop committed:
доб

Dedtop committed:
пофиксел краш

Dedtop committed:
попровка

Dedtop committed:
проверка

Dedtop committed:
поправка цвета

Dedtop committed:
правка цвета

Dedtop committed:
еще раз правил цвета

Dedtop committed:
поправка

Dedtop committed:
еще раз правил цвет

Dedtop committed:
поправил все

Dedtop committed:
закруглил фотки

Dedtop committed:
поправил настройки

Dedtop committed:
добавление кода и 2 экрана

Dedtop committed:
добавил в асет фотку выхода и добавил лейбал

Dedtop committed:
добавил кнопки

Dedtop committed:
назначил экран

Dedtop committed:
пофиксел экран

Dedtop committed:
1

Dedtop committed:
добавил таб бар и добавил иконки к нему

Dedtop committed:
добавление кода

Dedtop committed:
потерял экран так как не закомил его и всзал его же у автора

Dedtop committed:
забыл вернуть код который тоже не закомился

Dedtop committed:
Добавил код

Dedtop committed:
сделал верстку экрана и добавил цвета в ассетс

Dedtop committed:
добавил кнопку поделиться

Dedtop committed:
Добавил код Auth и добавил экран для него

Dedtop committed:
добавление кода и новых экранов

Dedtop committed:
добавления кода

Dedtop committed:
добавил

Dedtop committed:
добавил код

Dedtop committed:
изменил код

Dedtop committed:
пробовал решить проблему с авторизации

Dedtop committed:
Добавил экран и к нему код

Dedtop committed:
изменения

Dedtop committed:
добавление индификатора

Dedtop committed:
пофиксил все проблемы

Dedtop committed:
чуть-чуть поправил от коментов в AuthViewController

Dedtop committed:
Добавил пак и забыл закомментить изменения в AuthViewController

Dedtop committed:
Добавил файл ProgressHUD

Dedtop committed:
Сдедал верстку кодом приложение, закинул 2 Swift Package Manager, сделал запросы для экрана профиля и сделал работу в многопоточной среде

Dedtop committed:
Поставил индификаторы на Таб бар и на имадже лист

Dedtop committed:
удалил таб бар с навигатора

Dedtop committed:
добавил в SceneDelegate  использование Сплэш

Dedtop committed:
Добавил связь с таб баром

Dedtop committed:
Пофиксел экран таб бар

Dedtop committed:
тоже самое

Dedtop committed:
оставил комментарий

Dedtop committed:
no message

Dedtop committed:
фиксел ошибки

Dedtop committed:
Добавил анимацию, выход из профиля и ленту фотографий

Dedtop committed:
Убрал комментария

Dedtop committed:
Добавил экраны, вернул коментарии для облегчения, и пофиксел кнопку лайк

Dedtop committed:
добавил экран

Dedtop committed:
no message

Dedtop committed:
no message

Dedtop committed:
еще раз добавил мейн

Dedtop committed:
еще раз создал мейн и перекинул все туда

Dedtop committed:
поменял все так как меня не устраивал код

Dedtop committed:
.

Dedtop committed:
ImagesListService.fetchPhotosNextPage содержит implicit unwrap.
В методе changeLike сильно захвачен self. Исправил

Dedtop committed:
.

Dedtop committed:
В случае ошибки загрузки теперь явно сбрасывается self.currentTask = nil. Это позволит заново инициировать загрузку при следующем вызове fetchPhotosNextPage

Dedtop committed:
.

Dedtop committed:
исправиление ошибок и убратья page

Dedtop committed:
не залились исправления

Dedtop committed:
добавил изменения

Dedtop committed:
.

Dedtop committed:
no message

Dedtop committed:
создания теста и устранения ошибок

Dedtop committed:
Добавления файлов и изменения, исправления ошибок

Dedtop committed:
исправил одну ошибку

Dedtop committed:
исправил еще одну ошибку

Dedtop committed:
исправил пару ошибок и добавил ImageProtocol

Dedtop committed:
исправил несколько ошибок

Dedtop committed:
добавил индефикатор

Dedtop committed:
добавил обложку и поправил код

Dedtop committed:
2 теста UI починил все работает

Dedtop committed:
Не сохранился индефикатор

Dedtop committed:
no message

Dedtop committed:
no message

Dedtop committed:
изменил пароль

Dedtop committed:
no message

Dedtop committed:
поменял тест

Dedtop committed:
no message

andrew-kuzn-yndx commented on ImageFeed/AlertManager/DateManager.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+class DateFormatterManager {
```
Тут тоже стоит указывать final, а не только у вью контроллеров
Рекомендую добавлять этот модификатор ко всем классам при их создании, а затем, если вдруг понадобится наследоваться от какого-то класса, то компилятор подскажет, что нужно убрать final)

andrew-kuzn-yndx commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    var presenter: WebViewPresenterProtocol?
+    weak var delegate: WebViewViewControllerDelegate?
+    
+    @IBOutlet private var webView: WKWebView!
```
Молодец, что указал private для аутлетов! Также хорошей практикой является указание модификатора weak к IBOutlet. Вероятность retain cycle тут мала, но и нам ничего не стоит написать слабые ссылки)

andrew-kuzn-yndx commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,95 @@
+        return .darkContent
+    }
+    
+    @IBAction private func didTapBackButton(_ sender: Any?) {
```
Можно использовать комментарии // MARK: - ... для организации кода.
Подробнее можно посмотреть в [этой статье](https://habr.com/ru/articles/744470/): 

Тут, например, можно разделить методы на приватные и публичные блоки

andrew-kuzn-yndx commented on ImageFeed/AuthHelper/AuthHelper.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    func code(from url: URL) -> String? {
+        if let urlComponents = URLComponents(string: url.absoluteString),
+           urlComponents.path == "/oauth/authorize/native",
```
Давай строки path тоже поместим в файл Configuration, который у тебя есть) используй его по полной

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,51 @@
+        likeButton.setImage(likeImage, for: .normal)
+    }
+    
+    @IBAction func tapLikeButton(_ sender: UIButton) {
```
Не забывай, что IBAction **всегда** должны быть приватными

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,160 @@
+}
+
+private func makeRequest(authToken: String, page: Int) -> URLRequest? {
+    guard var urlComponents = URLComponents(string: "https://api.unsplash.com/photos") else {
```
Этот url тоже можно в Configuration

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,160 @@
+}
+
+private func likeRequest(photoId: String, isLike: Bool) -> URLRequest? {
+    let baseURL = "https://api.unsplash.com"
```
а этот url стоит брать из Configuration

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesViewController.swift:
```diff
@@ -0,0 +1,140 @@
+                
+            case .failure(let error):
+                self.showLikeErrorAlert(with: error)
+                UIBlockingProgressHUD.dismiss()
```
Дублирование UIBlockingProgressHUD. Его можно вывести за пределы switch, так как его вызывать в любом случае)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,131 @@
+        loadAndShowImage(url: image)
+    }
+    
+    func loadAndShowImage(url: URL?) {
```
Этот метод не используется нигде, кроме этого класса, его можно инкапсулировать
И методы ниже - тоже

andrew-kuzn-yndx commented on ImageFeed/ImagesList/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,131 @@
+        present(share, animated: true, completion: nil)
+    }
+    @IBOutlet private var imageView: UIImageView!
+    @IBAction private func didTapBackButton(_ sender: UIButton) {
```
Обычно используем такой порядок: сначала перечисляем свойства класса, затем функции. 
При этом также делим на приватные и публичные блоки. 
То есть твои IBAction должны быть выше, с другими свойствами

andrew-kuzn-yndx commented on ImageFeed/OAuth/OAuth2Service.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+    
+    private func authTokenRequest(code: String) -> URLRequest? {
+        guard var urlComponents = URLComponents(string: "https://unsplash.com/oauth/token") else {
```
Пройдись по всему проекту на предмет свободно лежащих ссылок
У тебя есть для этого отдельный инструмент, пусть ссылки там хранятся

andrew-kuzn-yndx commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,140 @@
+        exitButton.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(exitButton)
+        
+        nameLabel.text = "Екатерина Новикова"
```
Для создания и настройки элемента, ты можешь воспользоваться следующей конструкцией, которая помогает настраивать элемент при его объявлении и не создавать дополнительные методы по настройке:
```
    private lazy var nameLabel: UILabel = {
        let label = UILabel()
        label.textColor = UIColor(named: "YP White")
         ...реализация и настройка
        return label
    }()
```

Вот тут в середине статьи есть другие примеры: https://dzen.ru/a/YmMsexP4yGGM3uf1

andrew-kuzn-yndx commented on ImageFeed/Profile/ProfileViewPresenter.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+    
+    func viewDidLoad() {
+        profileImageServiceObserver = NotificationCenter.default
```
Давай тут тоже добавим удаление observer'a, как ты это сделал в другом классе

andrew-kuzn-yndx commented on ImageFeed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+        
+        @IBOutlet private var imageView: UIImageView!
```
Кажется, тут у тебя съехали отступы.
Отступы помогают коду выглядеть аккуратнее и привычнее взгляду, так читать код намного удобнее. Чтобы поправить оступы: можно выделить весь код cmd+A и зажать contol+i

andrew-kuzn-yndx commented on ImageFeed/Splash/SplashViewController.swift:
```diff
@@ -0,0 +1,126 @@
+                UIBlockingProgressHUD.dismiss()
+                
+            case .failure:
+                UIBlockingProgressHUD.dismiss()
```
И здесь дублирование UIBlockingProgressHUD

Dedtop committed:
1. В тесте testAuth добавлен свайп для скрытия клавиатуры.
2. В тесте testFeed убеждаюсь, что бесконечный вызов fetchPhotosNextPage предотвращён.
3. Исправлена ошибка передачи типа Error.self в методе showLikeErrorAlert.

Dedtop committed:
добавил индефикатор,
 - Добавлена проверка существования элементов ScrollView и изображения перед тем, как выполнить с ними действия. Используются waitForExistence и XCTAssertTrue для проверки доступности этих элементов.
- Применение жеста pinch к изображению осуществляется только после того, как элементы найдены

Dedtop committed:
Добавил кнопке назад индефикатор

Dedtop committed:
добавил индефикатор

Dedtop committed:
не докоминдел индефикатор

Dedtop committed:
Отключение пагинации в режиме тестирования,Идентификатор для кнопки "Back",Обработка кнопки "Like",Зум изображения

Dedtop committed:
Забыл коминтить в тот коминт

Dedtop committed:
1. Добавлен метод `isRunningUITest()`, который проверяет наличие аргумента UITests. Этот метод используется в tableView(_:willDisplay:forRowAt:), чтобы не выполнять пагинацию при тестировании.
2. В тестовом режиме (при наличии аргумента `"UITests"`) метод пагинация отключается.

Dedtop committed:
не закинул в теже изменения

Dedtop committed:
1. Добавлено ожидание: После нажатия кнопки логаута добавлено ожидание появления предупреждения. Это позволяет удостовериться, что экран логаута был полностью открыт перед дальнейшими действиями.2.- Добавлено ожидание после нажатия кнопок "Like" и "Unlike".
     - Добавлено ожидание при открытии изображения и перед его увеличением и уменьшением.3. Инициализация Presenter: Инициализация presenter перемещена в viewDidLoad, чтобы гарантировать, что она выполняется, когда view полностью загружено. Это предотвращает возможные проблемы с доступом к nil-значениям.Безопасные проверки: Мы добавили проверки на nil для tableView, чтобы убедиться, что она инициализирована. Тесты: Тест проверяет, что presenter инициализирован, что должно предотвратить сбои.

Dedtop committed:
- testProfileViewControllerUpdateAvatar проверяет, был ли вызван метод updateAvatar.
    - testImagesViewControllerCallsViewDidLoad проверяет, был ли установлен presenter после загрузки view.
    - testProfileViewControllerUpdateProfile проверяет, был ли вызван метод updateProfileDetails в презентере.

Dedtop committed:
добавил Presenter

Dedtop committed:
попарвил ошибки и добавил ImagSpay

Dedtop committed:
добавил мок и поправил Image test

Dedtop committed:
поправил testImagesViewControllerCallsViewDidLoad

Dedtop committed:
не закинулось изменение

Dedtop committed:
и это тоже

Dedtop committed:
не сохранилось все  изменения в testImagesViewControllerCallsViewDidLoad

Dedtop committed:
не прогружались фотки из-за изменений имадж сервесе

Dedtop committed:
поправил тест UI TestFeed

Dedtop committed:
не закинулось в то изменение

Dedtop committed:
случайно был удален testAuth

Dedtop committed:
достал прошлый UI

Dedtop committed:
нашел свои ошибки в тесте UI

Dedtop committed:
добавил проверку в презентр, что если ты в состоянии тестов ты загружаешь только первую страницу

Dedtop committed:
не правильно написал
